url10fb6bd43e_guwere_P2PSearch_tgz-pJ8-MainJ8/src/Controller.java

```diff
@@ -0,0 +1,577 @@
+		if(routingTable.containsKey(sender_id)){
+			routingTable.get(sender_id).setTimestamp(currTime);
+		}else{
+			routingTable.put(sender_id, new Contact(sender_id, "127.0.0.1"));
+		}
+	}
+    private void updateRouteTable(JSONArray table){
+    	for(Object row: table){
+    		JSONObject tableNode = (JSONObject) row;
+    		int nodeID = Integer.valueOf((String)tableNode.get("node_id"));
+    		Contact tableContact = new Contact(nodeID, (String)tableNode.get("ip_addresss"));
+    		tableContact.setTimestamp(System.currentTimeMillis());
+    		routingTable.put(nodeID,tableContact);
+    	}
+    }
+    /*
+    * Put an internal periodic update routing table message for the controller to handle.
+    * Delegated to the Controller rather than here in the run method to prevent mutual exclusion
+    * problems on the routing table.Otherwise the table has to be locked if there is more than
+    * one point of entry to it.
+    * UPDATE: the rule no longer applies as the routingTable was changed 
+    * to ConcurrentHashMap so different threads do not have to worry about the table consistency.
+    * */
+    private class RefreshRoutingTable extends TimerTask {
+        @Override
+        public void run() {
+            try {
+                JSONObject message =  new JSONObject();
+                message.put("type","REFRESH_ROUTING_TABLE");
+                receiveQueue.put(jsonToDatagram(message,new InetSocketAddress("",0)));
+            } catch (InterruptedException e) {
+                System.err.println("Could not put internal refresh table message in Controller");
+                e.printStackTrace();
+            }
+        }
+    }
+    
+    private class Waiter {
+    	Map<String,Timer> wordTimers;// <word, timer for word>
+    	public boolean transactionFinalized() {
+    		return wordTimers.isEmpty();
+    	}
+        public Waiter(){
+            wordTimers = new HashMap<String, Timer>();
+        }
+    	public void stopWait(String word) {
+    		wordTimers.get(word).cancel();
+    		wordTimers.remove(word);    		
+    	}
+    	// the wait for a particular word will be cancelled with the stopWait method
+    	public void startWait(String word, final int targetHash) {
+    		Timer responseTimer = new Timer();
+    		responseTimer.schedule(new TimerTask() {
+    			@SuppressWarnings("unchecked")
+    			  public void run() {//the run method will be run only after the scheduled time has expired
+    				  	//send a ping to target through the closest node.Might even be the actual node
+    				  	Contact closestNode = getClosestNode(targetHash);
+    				  	if(closestNode  != null){
+    				  		JSONObject message = getPingJson(closestNode);
+        				  	InetSocketAddress address = new InetSocketAddress(closestNode.getIp(),NetworkProperties.nPort);           				
+        			        try {
+        			        	sendQueue.put(jsonToDatagram(message,address));
+        			        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)